### ISSUE TITLE
Retina support

### ISSUE THREAD

**Issue by lgarron** [cubing/AnimCubeJS#2]
![screenshot 2015-06-29 23 16 21](https://cloud.githubusercontent.com/assets/248078/8424882/e273c250-1eb4-11e5-8b6a-7fb8cc19dab3.png)

See `devicePixelRatio` and `[webkit]BackingStorePixelRatio`.

**Comment by lgarron** [comment]
It seems all the canvases are constructed in the obfuscated GWT output. I can't find the sources for those, which would make it very not semantically awesome for me to tinker with this.

**Comment by bcube2** [comment]
I have no idea what are you talking about, but do you think Michael as an author could help? Have you tried to contact him?

**Comment by molarmanful** [comment]
@bcube2 Lucas is talking about adding some code to make the cubes render better in devices with retina display.

@lgarron I'm also working on it, but my approach is to inject a separate script to deal with the canvas tags. No idea if that will work, but I'm trying it out.

**Comment by lgarron** [comment]
Also, would you like this at `https://github.com/cubing/AnimCubeJS`?
Either way is fine, but this definitely qualifies as a valuable community project.

**Comment by molarmanful** [comment]
Sure! That would be great.

**Comment by molarmanful** [comment]
@lgarron Okay, I've almost got the retina display support down, except that my injected script runs _before_ the canvas loads, which throws an error... I'll need some help with fixing this (see cube.html for my script).

**Comment by lgarron** [comment]
You should now be a (temporary) admin of the `cubing` org. YOu should be able to transfer now.

**Comment by molarmanful** [comment]
Hmm... I don't seem to have transfer permissions. Are you sure you gave me permissions?

**Comment by lgarron** [comment]
It seems you have to accept an invitation.

**Comment by molarmanful** [comment]
Okay, now that I'm in, is it okay for me to add @blonkm and @bcube2 as collaborators on the AnimCubeJS team?

**Comment by lgarron** [comment]
Yeah, go ahead; you're in charge of the project. :-)

**Comment by molarmanful** [comment]
@lgarron Okay, Michael Feather has released unminified versions of the Javascript code (see readme for link). No idea if that will help at all, but at least it's better to work with.

By the way, I deleted the previous repo transfer conversation to get us back on track.

**Comment by mfeather1** [comment]
I'm not exactly sure what you are trying to do here but from looking at cube.html it looks like the goal is to multiply canvas width & height by  
window.devicePixelRatio when it's  greater than 1.  However, on the demo page the iframes that contain the cubes have height="100%" width="100% so they already fill the width of whatever size window they are displayed in (at least that's what happens with Firefox that I'm using) so they could not be scaled any wider.  If iframe width & height are not specified (nor specified as parameters to cube.html) then the cube window will be the size of whatever it's contained in, which you can see if you access AnimCube.html (on my site) directly -- it takes up the whole browser window.  So the size of the cube can be controlled by 1) setting the size of the iframe or 2) setting the size of whatever the iframe is contained in (using 100% for iframe params) or 3) passing the size as parameters directly to AnimCube.html (although that would have to be in pixels).

**Comment by mfeather1** [comment]
try this: 
http://www.irbsystems.com/test2/retina/test.html

resizing works if I hardcode scaleFactor = 2
(I don't have a touch device to test with)

**Comment by mfeather1** [comment]
I tested this on a touch device and it works.

**Comment by molarmanful** [comment]
Great! That's done...

**Comment by lgarron** [comment]
That's... interesting.

<img width="1392" alt="screenshot 2015-07-22 01 06 56" src="https://cloud.githubusercontent.com/assets/248078/8820880/ff6591da-300d-11e5-86c4-8da54af18132.png">

(Also, the resolution doesn't change if you drag the window from a regular screen to a Retina screen, but I guess that's to be expected if you don't want to get fiddly with CSS.)

**Comment by bcube2** [comment]
"Lucas is talking about adding some code to make the cubes render better in devices with retina display."

By that you mean the cubes will get bigger in devices with retina display? Because somehow I got bigger cubes (as shown on the picture above) on mobile phone where I have a fixed width of a website, thus it doesn´t look as good as previous version without "retina enhancement".

So do I need the old cube.html where it was OK on mobile device?

**Comment by bcube2** [comment]
Edit: I got back to the old version of cube.html on my website, until some of you guys answers/helps me.

**Comment by mfeather1** [comment]
I thought the issue was to scale the image up by a factor of devicePixelRatio but apparently the touch devices do that already (my lack of retina device makes it a bit difficult to work on this, the one I had (which was borrowed and I just returned) only had a  devicePixelRatio of 1.3).  What is needed is a way to tell the device (via JS) to not auto-upsize the image and just render the cube directly into the screen pixels (retina's 2 to 1 ratio of physical pixels to logical pixels would then scale down the image properly).  I could not find a way to do that so I will take the resizing code out of cube.html and reopen this issue.

**Comment by mfeather1** [comment]
Closing this, will re-open when I have a retina device to test with.

**Comment by bcube2** [comment]
On http://www.irbsystems.com/test2/retina/test.html on mobile-phone, I observe strange behaviour of the first simulator.

When clicked "to the end" or "one step forward" button, it ends in RUF position (so far so good). When clicked "play" button, it does R, then U and then it resets the U move and do U move again, so it ends with RU position only.

Possible source of this error: ww=1 parameter (no idea what Michael wanted to do...).

**Comment by mfeather1** [comment]
The purpose of that test page is just to see if there is any difference in the image quality between the two cubes displayed.  If the code does what it's supposed to then the second cube should look better on a retina display but I would not be surprised if it doesn't because lacking a way to test it's not likely that I got it right with one try.  To fix the bug you found (with my undocumented ww param) I took out the move sequence as it's not needed.  I could probably say more about it but since that param isn't likely to be used by anyone other than me, I'm going to leave it at that, if you want to know more about it, say so and I'll expound...
